### PR TITLE
Basic reconstruction and discrimination games

### PR DESCRIPTION
**Pull request by mbaroni**
##  Description
I added basic reconstruction and discrimination games reading input from text files, and with detailed inline comments. They will hopefully be useful as starting points to understand and play with EGG for people with my kind of non-purely computational background.

## Motivation and Context
While there are a lot of interesting games in EGG, we don't really have basic implementations of the most basic games. Moreover, we don't have much code with detailed comments-based documentation..

## How Has This Been Tested?
I tested the games with small datasets, considering various limit cases.
I also reproduced previous results from early experiments with other code when using the same hyperparameters.
I added a basic test for pytest (and checked that the code passes it).

### CODE REVIEW

File: tests/test_games_do_not_fail.py [hunk 1]
```diff
@@ -44,6 +44,21 @@ def test_objects_game():
         ),
     )
 
+def test_basic_games_game():
+    run_game(
+        "egg.zoo.basic_games.play",
+        dict(
+            mode="gs",
+            train_data="egg/zoo/basic_games/data_generation_scripts/example_reconstruction_input.txt",
+            validation_data="egg/zoo/basic_games/data_generation_scripts/example_reconstruction_input.txt",
```
Comment (robertodessi) [inline tests/test_games_do_not_fail.py:53]: any reason for not testing the discrimination game? I guess they share most of the code so maybe not really worth it, right?

File: egg/core/language_analysis.py [hunk 2]
```diff
@@ -260,4 +260,34 @@ def on_epoch_end(self, _loss, logs: Interaction, epoch: int):
...
+        print([m.tolist() for m in logs.message], sep='\n')
+        print("OUTPUTS")
+        print([m.tolist() for m in logs.receiver_output], sep='\n')
+        
+    def on_test_end(self, _loss, logs: Interaction, epoch: int): # here is where we make sure we are printing the validation set (on_test_end, not on_epoch_end)
+        if (epoch==self.n_epochs): # here is where we check that we are at the last epoch
+            self.print_events(logs)
+
+    def on_early_stopping(self,  _train_loss, _train_logs, epoch, _test_loss, test_logs): # same behaviour if we reached early stopping
+        self.print_events(test_logs)
```
Comment (robertodessi) [inline egg/core/language_analysis.py:292]: Was this tested? I thing the signature should be:

```
def on_early_stopping(
        self,
        train_loss: float,
        train_logs: Interaction,
        epoch: int,
        test_loss: float = None,
        test_logs: Interaction = None,
    ):
```

Comment (mbaroni) [reply egg/core/language_analysis.py:292]: I have tested it and it works. And indeed it seems to me to have the correct arguments: where do you see a discrepancy?

File: tests/test_games_do_not_fail.py (hunk 1 above)
Comment (mbaroni) [reply tests/test_games_do_not_fail.py:53]: The games are very similar, and testing is a bit slow, so I would just test one.

**Pull request merged**